design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,114 @@
+# Uncommitted Garbage Collector
+
+## Motivation
+
+Uncommitted data which is no longer referenced (due to branch deletion, reset branch etc.) is not being deleted by lakeFS.
+This may result excessive storage usage and possible compliance issues.
+To solve this problem two approaches were suggested:
+1. A batch operation performed as part of an external process (GC)
+2. An online solution inside lakeFS
+
+Several attempts for an online solution have been made, most of which are documented [here](Link-to-hard-delete-proposal).
+This document will describe the **offline** GC process for uncommitted objects.
+
+Garbage collection of uncommitted data will be performed using the same principals of the current GC process.
+The basis for this is a GC Client (i.e. _Spark_ job) consuming objects information from both lakeFS and the storage adapter, 
+and using this information to determine which objects on the bucket can be deleted.
+
+For simplicity, in this document we will reduce this problem to the repository and branch level.
+
+## Design
+
+### Required changes by lakeFS
+
+The following are necessary changes in lakeFS in order to implement this proposal successfully.
+
+#### Objects Path Conventions
+
+1. Repository objects will be stored under the prefix `<bucket_name>/repos/<repo_uid>`
+2. Branch objects will be stored under the repo prefix with the path `branches/<branch_id>/`
+3. Each lakeFS instance will create a unique prefix partition (serialized) under the branch path to 
```

Comment:
Effectively it will look the same.
Each lakeFS instance will maintain it's own partitions appending a unique suffix to the partition name, in the following manner:
`<sortable_ascending_serialized_uid>_<lakefs_instance_uid>`
Each instance will manage an exclusive partition  - on container restart the lakeFS instance will created a new partition according to the above conventions.
The current idea is to track partition size in memory - this means a restarted instance does not have context on the existing partitions, effectively leaving partitions partially allocated. If you think this is not sufficient we might want to think of an additional solution on how to track the partitions

Reply:
Updated document